File: chapitre-25/exemple-select.py
#! /usr/bin/python
# ------------------------------------------------------------------
# exemple-select.py
# Fichier d'exemple du livre "Developpement Systeme sous Linux"
# (C) 2015-2019 - Christophe BLAESS <christophe@blaess.fr>
# https://www.blaess.fr/christophe/
# ------------------------------------------------------------------

from __future__ import print_function
import os
import fcntl
import select
import sys
import time

read_list = []

for i in range (0, 9):
	(fd_r, fd_w) = os.pipe()
	if os.fork() != 0:
		os.close(fd_r)
		while True:
			time.sleep(i+1)
			os.write(fd_w, chr(48+i))
	else:
		read_list.append(fd_r)

while True:
	(r, w, x) = select.select(read_list, [], [])
	for fd in r:
		c = os.read(fd, 4096)
		print(c, end=' ', file=sys.stderr)

File: chapitre-25/exemple-select.c
// ------------------------------------------------------------------
// exemple-select.c
// Fichier d'exemple du livre "Developpement Systeme sous Linux"
// (C) 2000-2019 - Christophe BLAESS <christophe@blaess.fr>
// https://www.blaess.fr/christophe/
// ------------------------------------------------------------------

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>

#define	CHILDS	10

int main (void)
{
	int pipe_fd[CHILDS][2];
	int i;
	int max = -1;
	char c;
	fd_set set;

	for (i = 0; i < CHILDS; i ++) {
		if (pipe (pipe_fd[i]) < 0) {
			perror("pipe");
			exit(EXIT_FAILURE);
		}
		if (fork() == 0) {
			// Enfant
			close(pipe_fd[i][0]);
			c = '0' + i;
			while(1) {
				sleep(i + 1);
				write(pipe_fd[i][1], & c, 1);
			}
		}
		close(pipe_fd[i][1]);
	}
	// Parent
	while (1) {
		FD_ZERO(& set);
		for (i = 0; i < CHILDS; i ++) {
			FD_SET(pipe_fd[i][0], & set);
			if (pipe_fd[i][0] > max)
				max = pipe_fd[i][0];
		}
		if (select(max, & set, NULL, NULL, NULL) <= 0) {
			perror("select");
			break;
		}
		for (i = 0; i < CHILDS; i ++) {
		 	if (FD_ISSET(pipe_fd[i][0], & set)) {
				read(pipe_fd[i][0], & c, 1);
				fprintf(stderr, "%d ", i);
			}
		}
	}
	return EXIT_SUCCESS;
}

